File: src/main/java/com/dongfg/project/api/util/RamRateLimiter.java
package com.dongfg.project.api.util;

import com.google.common.util.concurrent.RateLimiter;

import java.util.HashMap;
import java.util.Map;

/**
 * @author dongfg
 * @date 2017/10/15
 */
public class RamRateLimiter {

    public static final String LIMIT_KEY_SMS = "sms";

    private static final Map<String, RateLimiter> RATE_LIMITER_MAP = new HashMap<>();

    static {
        RATE_LIMITER_MAP.put(LIMIT_KEY_SMS, RateLimiter.create(0.1));
    }

    public static boolean acquire(String limitKey) {
        RateLimiter rateLimiter = RATE_LIMITER_MAP.get(limitKey);
        return rateLimiter.tryAcquire();
    }

    private RamRateLimiter() {

    }
}

File: src/main/kotlin/com/dongfg/project/api/common/util/RamRateLimiter.kt
package com.dongfg.project.api.common.util

import com.dongfg.project.api.common.Constants.RateLimitKey
import com.google.common.util.concurrent.RateLimiter
import java.util.concurrent.ConcurrentHashMap

class RamRateLimiter {
    companion object {
        private val limiterMap = ConcurrentHashMap<RateLimitKey, RateLimiter>()

        init {
            limiterMap[RateLimitKey.MESSAGE] = RateLimiter.create(0.1)
        }

        @JvmStatic
        fun acquire(rateLimitKey: RateLimitKey): Boolean {
            return limiterMap[rateLimitKey]!!.tryAcquire()
        }
    }
}